# Analiza kolumny system: details, resources, saves, traits


Import danych

In [2]:
from analysis_functions import (
    import_merged_bestiaries,
    unpack_column,
    remove_redundant_level,
)
import pandas as pd


bestiary = import_merged_bestiaries(
    [
        "../../pathfinder_2e_data/pathfinder-bestiary.db",
        "../../pathfinder_2e_data/pathfinder-bestiary-2.db",
        "../../pathfinder_2e_data/pathfinder-bestiary-3.db",
    ]
)

In [3]:
system = unpack_column(bestiary, "system")

In [4]:
system.head()

abilities  \
0  {'cha': {'mod': 1}, 'con': {'mod': 5}, 'dex': ...   
1  {'cha': {'mod': 1}, 'con': {'mod': 5}, 'dex': ...   
2  {'cha': {'mod': -2}, 'con': {'mod': 6}, 'dex':...   
3  {'cha': {'mod': 6}, 'con': {'mod': 5}, 'dex': ...   
4  {'cha': {'mod': 1}, 'con': {'mod': 1}, 'dex': ...   

                                          attributes  \
0  {'ac': {'details': '', 'value': 29}, 'allSaves...   
1  {'ac': {'details': '', 'value': 28}, 'allSaves...   
2  {'ac': {'details': '', 'value': 25}, 'allSaves...   
3  {'ac': {'details': '', 'value': 41}, 'allSaves...   
4  {'ac': {'details': '', 'value': 16}, 'allSaves...   

                                             details  \
0  {'alignment': {'value': 'CE'}, 'blurb': '', 'c...   
1  {'alignment': {'value': 'N'}, 'blurb': '', 'cr...   
2  {'alignment': {'value': 'N'}, 'blurb': '', 'cr...   
3  {'alignment': {'value': 'LE'}, 'blurb': '', 'c...   
4  {'alignment': {'value': 'LG'}, 'blurb': '', 'c...   

                           resources  \
0  {'focus': {'max': 1, 'value': 1}}   
1                                 {}   
2                                 {}   
3  {'focus': {'max': 1, 'value': 1}}   
4  {'focus': {'max': 1, 'value': 1}}   

                                               saves  \
0  {'fortitude': {'saveDetail': '', 'value': 22},...   
1  {'fortitude': {'saveDetail': '', 'value': 21},...   
2  {'fortitude': {'saveDetail': '', 'value': 20},...   
3  {'fortitude': {'saveDetail': '', 'value': 30},...   
4  {'fortitude': {'saveDetail': '', 'value': 6}, ...   

                                              traits  \
0  {'languages': {'custom': '', 'selected': [], '...   
1  {'languages': {'custom': '', 'selected': [], '...   
2  {'languages': {'custom': '', 'selected': [], '...   
3  {'languages': {'custom': '', 'selected': [], '...   
4  {'languages': {'custom': '', 'selected': [], '...   

                                              schema  
0  {'version': 0.8270000000000001, 'lastMigration...  
1  {'version': 0.8270000000000001, 'lastMigration...  
2  {'version': 0.8270000000000001, 'lastMigration...  
3  {'version': 0.8270000000000001, 'lastMigration...  
4  {'version': 0.8270000000000001, 'lastMigration...

In [5]:
system.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   abilities   1206 non-null   object
 1   attributes  1206 non-null   object
 2   details     1206 non-null   object
 3   resources   1206 non-null   object
 4   saves       1206 non-null   object
 5   traits      1206 non-null   object
 6   schema      1206 non-null   object
dtypes: object(7)
memory usage: 75.4+ KB


## Details

In [6]:
details = unpack_column(system, "details")

In [7]:
details.head()

alignment blurb creatureType          level privateNotes  \
0  {'value': 'CE'}             Dragon  {'value': 10}                
1   {'value': 'N'}             Animal  {'value': 10}                
2   {'value': 'N'}             Animal   {'value': 8}                
3  {'value': 'LE'}             Dragon  {'value': 17}                
4  {'value': 'LG'}          Celestial   {'value': 1}                

                                         publicNotes  \
0  <p>The most feral and least intelligent of all...   
1  <p>Brontosauruses are truly gigantic behemoths...   
2  <p>The monstrous giant anaconda is capable of ...   
3  <p>Green dragons are the most contemplative of...   
4  <p>Friendly, curious, and eternally optimistic...   

                             source  
0  {'value': 'Pathfinder Bestiary'}  
1  {'value': 'Pathfinder Bestiary'}  
2  {'value': 'Pathfinder Bestiary'}  
3  {'value': 'Pathfinder Bestiary'}  
4  {'value': 'Pathfinder Bestiary'}

Jak widać wyżej mamy sporo fałszywych wartości non-null -> zmieniamy je na np.nan (null)

In [8]:
import numpy as np


details = details.replace("", np.nan)

In [9]:
details.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   alignment     1206 non-null   object 
 1   blurb         0 non-null      float64
 2   creatureType  825 non-null    object 
 3   level         1206 non-null   object 
 4   privateNotes  0 non-null      float64
 5   publicNotes   1206 non-null   object 
 6   source        1206 non-null   object 
dtypes: float64(2), object(5)
memory usage: 75.4+ KB


* **alignment** - do sprawdzenia
* blurb - i tak nic nie ma
* **creatureType** - do sprawdzenia
* **level** - "do sprawdzenia"
* privateNotes - i tak nic nie ma
* **publicNotes** - do sprawdzenia
* **source** - do sprawdzenia

### Details: Alignment

In [10]:
alignment = unpack_column(details, "alignment")

In [11]:
alignment.columns

Index(['value'], dtype='object')

In [12]:
alignment.head()

value
0    CE
1     N
2     N
3    LE
4    LG

In [13]:
alignment.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   value   1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


Same values -> można zrobić to co z mod

In [14]:
alignment.value.unique()

array(['CE', 'N', 'LE', 'LG', 'CG', 'NE', 'CN', 'NG', 'LN'], dtype=object)

Mamy 9 unikatowych wartości
 * czy są one istotne????
 * czy przydadzą się do if -ów???

### Details: CreatureType

In [15]:
types = unpack_column(details, "creatureType")

In [16]:
types.head()

0
0     Dragon
1     Animal
2     Animal
3     Dragon
4  Celestial

In [17]:
types.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       825 non-null    object
dtypes: object(1)
memory usage: 18.8+ KB


Z jakiegoś powodu nie każdy potworek ma typ:
 * czy to problem?
 * czy może być więcej?

In [18]:
all_types = types[0].unique()
all_types

array(['Dragon', 'Animal', 'Celestial', 'Aberration', 'Undead',
       'Humanoid', 'Construct', 'Spirit', 'Elemental', 'Fey', 'Beast',
       'Fiend', 'Plant', 'Giant', nan, 'Monitor', 'Ooze', 'Fungus',
       'Astral', 'Animal, Undead', 'Dragon, Elemental', 'Beast, Dragon',
       'Giant, Humanoid', 'Petitioner', 'Beast, Fiend', 'Fungus, Ooze',
       'Spirit, Undead', 'Fiend, Giant, Humanoid', 'Dragon, Plant',
       'Fey, Plant', 'Aberration, Time', 'Aberration, Negative',
       'Beast, Fey', 'Beast, Humanoid', 'Humanoid, Positive', 'Dream',
       'Dragon, Undead', 'Aberration, Ethereal', 'Aberration, Dream',
       'Fungus, Plant'], dtype=object)

In [25]:
types_dict = {}
for t_list in all_types:
    if type(t_list) == str:
        for t in t_list.split(", "):
            types_dict[t] = None

In [26]:
types_dict.keys()

dict_keys(['Dragon', 'Animal', 'Celestial', 'Aberration', 'Undead', 'Humanoid', 'Construct', 'Spirit', 'Elemental', 'Fey', 'Beast', 'Fiend', 'Plant', 'Giant', 'Monitor', 'Ooze', 'Fungus', 'Astral', 'Petitioner', 'Time', 'Negative', 'Positive', 'Dream', 'Ethereal'])

In [27]:
print(len(types_dict))

24


Mamy 24 typy (+ pusty) - **może się przydać później do tych podpowiedzi if-owych**
*CR: I tak, do tych podpowiedzi się przyda*

### Datails: Level

In [28]:
level = unpack_column(details, "level")

In [29]:
level.head()

value
0     10
1     10
2      8
3     17
4      1

In [30]:
level.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   value   1206 non-null   int64
dtypes: int64(1)
memory usage: 18.8 KB


In [34]:
level.value_counts()

value
 1       127
 2       121
 3        95
 4        82
 7        77
 5        76
 6        72
 8        68
 9        63
 10       58
 12       41
 11       40
 13       36
-1        34
 14       33
 15       31
 0        27
 16       25
 20       24
 17       21
 18       21
 19       18
 21        8
 23        4
 24        2
 22        1
 25        1
dtype: int64

# **Najważniejsza wartość! - to co** przwidujemy
analogicznie jak abilities - mod

### Details: PublicNotes

In [21]:
notes = unpack_column(details, "publicNotes")

In [22]:
notes.head()

0
0  <p>The most feral and least intelligent of all...
1  <p>Brontosauruses are truly gigantic behemoths...
2  <p>The monstrous giant anaconda is capable of ...
3  <p>Green dragons are the most contemplative of...
4  <p>Friendly, curious, and eternally optimistic...

In [23]:
notes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


Zwykłe opisy do modelu raczej nie są potrzebne

### Details: Source

In [24]:
source = unpack_column(details, "source")

In [25]:
source.head()

value
0  Pathfinder Bestiary
1  Pathfinder Bestiary
2  Pathfinder Bestiary
3  Pathfinder Bestiary
4  Pathfinder Bestiary

In [26]:
source.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   value   1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


In [27]:
source.value.unique()

array(['Pathfinder Bestiary', 'Pathfinder Core Rulebook',
       'Pathfinder Bestiary 2', 'Pathfinder Bestiary 3'], dtype=object)

Skąd pochodzą potworki

### Details: podsumowanie
* **alignment**
    * do rozpakowania (abilities:mod case)
    * każdy potworek ma
    * czy są istotne?
    * może się przydać do if-ów
    * czy są istotne?
* blurb
    * puste
* **creatureType**
    * nie każdy potworek ma
    * może się przydać do if-ów
* **level**
    * do rozpakowania (abilities:mod case)
    * każdy potworek ma
    * CR
* privateNotes
    * puste
* **publicNotes**
    * każdy potworek ma
    * opis potworków
    * zbędne
* **source**
    * każdy potworek ma
    * książka z której jest potworek
    * raczej zbędne
    * wydaje mi się że było polecenie, żeby zostawić (nie do samego przewidywania, ale podizału na zbiory)

## Resources

In [35]:
system.resources

0      {'focus': {'max': 1, 'value': 1}}
1                                     {}
2                                     {}
3      {'focus': {'max': 1, 'value': 1}}
4      {'focus': {'max': 1, 'value': 1}}
                     ...                
360    {'focus': {'max': 0, 'value': 0}}
361    {'focus': {'max': 0, 'value': 0}}
362                                   {}
363    {'focus': {'max': 0, 'value': 0}}
364    {'focus': {'max': 1, 'value': 1}}
Name: resources, Length: 1206, dtype: object

In [69]:
system.resources = system.resources.replace([{}], [np.nan])

In [70]:
system.resources

0      {'focus': {'max': 1, 'value': 1}}
1                                    NaN
2                                    NaN
3      {'focus': {'max': 1, 'value': 1}}
4      {'focus': {'max': 1, 'value': 1}}
                     ...                
360    {'focus': {'max': 0, 'value': 0}}
361    {'focus': {'max': 0, 'value': 0}}
362                                  NaN
363    {'focus': {'max': 0, 'value': 0}}
364    {'focus': {'max': 1, 'value': 1}}
Name: resources, Length: 1206, dtype: object

In [73]:
system_not_null_resources = system[system.resources.notnull()]

In [77]:
print(len(system_not_null_resources))

566


In [78]:
print(len(system[system.resources.isnull()]))

640


In [74]:
resources = unpack_column(system_not_null_resources, "resources")

In [75]:
resources.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 566 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   focus   566 non-null    object
dtypes: object(1)
memory usage: 8.8+ KB


In [76]:
resources.head()

focus
0  {'max': 1, 'value': 1}
3  {'max': 1, 'value': 1}
4  {'max': 1, 'value': 1}
5  {'max': 1, 'value': 1}
6  {'max': 1, 'value': 1}

In [79]:
def check_diff_values(df, values):
    return (
        len(
            [
                _
                for i, row in df.iterrows()
                for col in df.columns
                for val in values
                if type(row[col]) is dict
                and (len(row[col].keys()) != len(values) or val not in row[col].keys())
            ]
        )
        > 0
    )

In [80]:
print("Klucze inne niż: max, value?")
print(check_diff_values(pd.DataFrame(data=resources.focus), ["max", "value"]))

Klucze inne niż: max, value?
False


In [81]:
max_val = {}

for i, row in resources.iterrows():
    if type(row["focus"]) is dict:
        max_val[(row["focus"].get("max"), row["focus"].get("value"))] = None

In [82]:
print(max_val)

{(1, 1): None, (0, 0): None, (3, 3): None, (2, 2): None}


### Resources
Zatem w bestiariuszach focus value == focus max

Zakres wartości: 0 - 3

Nie każdy potworek ma

Można zredukować złożoneość tej kolumny - ma 1 "pod kolumne" z 2 wartościami, które są takie same dla danego potworka o
ile je ma - **może się przydać**

*CR: Taką mam myśl, że jak potwór jest smokiem to będzie znał draconic. Ale to chyba bardziej nam jest potrzebna informacja że jest smokiem do tych wskazówek.*

## Saves

In [43]:
saves = unpack_column(system, "saves")

In [44]:
saves.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   fortitude  1206 non-null   object
 1   reflex     1206 non-null   object
 2   will       1206 non-null   object
dtypes: object(3)
memory usage: 37.7+ KB


In [45]:
saves.head()

fortitude                           reflex  \
0  {'saveDetail': '', 'value': 22}  {'saveDetail': '', 'value': 19}   
1  {'saveDetail': '', 'value': 21}  {'saveDetail': '', 'value': 14}   
2  {'saveDetail': '', 'value': 20}  {'saveDetail': '', 'value': 17}   
3  {'saveDetail': '', 'value': 30}  {'saveDetail': '', 'value': 29}   
4   {'saveDetail': '', 'value': 6}  {'saveDetail': '', 'value': 10}   

                              will  
0  {'saveDetail': '', 'value': 17}  
1  {'saveDetail': '', 'value': 16}  
2  {'saveDetail': '', 'value': 15}  
3  {'saveDetail': '', 'value': 32}  
4   {'saveDetail': '', 'value': 4}

Dane wyglądają "przyjaźnie", tylko jeszcze trzeba sprawdzić "saveDetails"

In [46]:
fortitude = unpack_column(saves, "fortitude")
reflex = unpack_column(saves, "reflex")
will = unpack_column(saves, "will")

### Saves - fortitude

In [47]:
fortitude.saveDetail.unique()

array(['', '(+8 vs. poisons)'], dtype=object)

In [48]:
fortitude[fortitude.saveDetail != ""]

saveDetail  value
334  (+8 vs. poisons)      6

tylko 1 potworek z jakąś wartością

In [49]:
fortitude.value.min()

0

In [50]:
fortitude.value.max()

44

In [51]:
fortitude.value.unique()

array([22, 21, 20, 30,  6, 15,  9, 13, 12,  8,  2, 10, 36, 35, 34, 16, 14,
       11, 27, 17,  7, 32, 28, 18, 23, 42,  5,  4, 26, 39, 25, 24,  3, 19,
       31, 29, 33, 37, 38,  0, 40, 44, 43, 41], dtype=int64)

In [52]:
for i in range(fortitude.value.min(), fortitude.value.max() + 1):
    if fortitude[fortitude.value == i].size < 1:
        print(i)

1


### Saves - reflex

In [53]:
reflex.saveDetail.unique()

array([''], dtype=object)

Brak detali

In [54]:
reflex.value.max()

41

In [55]:
reflex.value.min()

0

In [56]:
reflex.value.unique()

array([19, 14, 17, 29, 10, 13, 15,  9,  0,  8,  7, 11, 32, 30, 28, 26,  5,
       18, 21, 16, 40, 25,  6, 22, 23, 35, 20, 12, 27, 24,  1,  3, 31, 34,
        4, 41, 33, 37, 39, 38, 36], dtype=int64)

In [57]:
for i in range(reflex.value.min(), reflex.value.max() + 1):
    if reflex[reflex.value == i].size < 1:
        print(i)

2


### Saves - will

In [58]:
will.saveDetail.unique()

array(['', '; +1 status to all saves vs. magic'], dtype=object)

In [59]:
will[will.saveDetail != ""]

saveDetail  value
40   ; +1 status to all saves vs. magic     19
142  ; +1 status to all saves vs. magic     10
315  ; +1 status to all saves vs. magic      8

Sporadyczne detale, ale i tak wszystkie takie same

In [60]:
will.value.min()

0

In [61]:
will.value.max()

43

In [62]:
will.value.unique()

array([17, 16, 15, 32,  4, 11, 14,  2,  5,  6,  3,  8, 31, 35, 25,  9, 12,
       22, 33, 21, 13, 18,  7, 10, 27, 43, 24, 26, 34, 23, 29, 19, 20, 28,
       30, 39, 38, 37,  0,  1, 40, 36, 42], dtype=int64)

In [63]:
for i in range(will.value.min(), will.value.max() + 1):
    if will[will.value == i].size < 1:
        print(i)

41


### Saves - podsumowanie
mamy trzy kolumny: fortitude, reflex, will

Wszystkie potworki mają nie puste values

Zakres wartości:
* fortitude: 0-44
* reflex: 0-41
* will: 0-43
Z jakiegoś powodu w każdej kolumnie istnieje dokładnie jadna wartość której nie ma żaden potworek (z powyższych przediałów)

**Wydaje mi się, że dane dość spoko do w miarę prostego dodania**
Czy są istotne?
*CR: Tak, to jest do uwzględnienia*

## Traits

In [75]:
traits = unpack_column(system, "traits")

In [78]:
traits.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   languages  1206 non-null   object
 1   rarity     1206 non-null   object
 2   senses     1206 non-null   object
 3   size       1206 non-null   object
 4   value      1206 non-null   object
 5   attitude   718 non-null    object
dtypes: object(6)
memory usage: 66.0+ KB


In [76]:
traits.head()

languages    rarity  \
0  {'custom': '', 'selected': [], 'value': ['comm...    common   
1        {'custom': '', 'selected': [], 'value': []}    common   
2        {'custom': '', 'selected': [], 'value': []}    common   
3  {'custom': '', 'selected': [], 'value': ['abys...  uncommon   
4  {'custom': '', 'selected': [], 'value': ['cele...    common   

                                              senses               size  \
0  {'value': 'darkvision, scent (imprecise) 60 fe...    {'value': 'lg'}   
1  {'value': 'low-light vision, scent (imprecise)...   {'value': 'grg'}   
2  {'value': 'low-light vision, scent (imprecise)...  {'value': 'huge'}   
3  {'value': 'darkvision, scent (imprecise) 60 fe...   {'value': 'grg'}   
4                            {'value': 'darkvision'}    {'value': 'sm'}   

                  value attitude  
0        [cold, dragon]      NaN  
1    [animal, dinosaur]      NaN  
2              [animal]      NaN  
3  [amphibious, dragon]      NaN  
4   [archon, celestial]      NaN

Mamy 6 kolumn:
* languages
* rarity
* senses
* size
* value
* attitude (wartości brakujące)

### Traits - languages

In [77]:
languages = unpack_column(traits, "languages")

In [79]:
languages.head()

custom selected                                              value
0              []                                 [common, draconic]
1              []                                                 []
2              []                                                 []
3              []  [abyssal, common, draconic, elven, jotun, sylvan]
4              []                    [celestial, draconic, infernal]

In [80]:
languages.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   custom    1206 non-null   object
 1   selected  1206 non-null   object
 2   value     1206 non-null   object
dtypes: object(3)
memory usage: 37.7+ KB


In [90]:
languages = languages.replace("", np.nan)

In [91]:
languages.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   custom    313 non-null    object
 1   selected  1206 non-null   object
 2   value     1206 non-null   object
dtypes: object(3)
memory usage: 37.7+ KB


In [92]:
languages.head()

custom selected                                              value
0    NaN       []                                 [common, draconic]
1    NaN       []                                                 []
2    NaN       []                                                 []
3    NaN       []  [abyssal, common, draconic, elven, jotun, sylvan]
4    NaN       []                    [celestial, draconic, infernal]

*CR: Raczej nie potrzebne informacje*
*Taką mam myśl, że jak potwór jest smokiem to będzie znał draconic. Ale to chyba bardziej nam jest potrzebna informacja że jest smokiem do tych wskazówek.*

### Traits - rarity

In [93]:
rarity = unpack_column(traits, "rarity")

In [94]:
rarity.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


In [95]:
rarity.head()

0
0    common
1    common
2    common
3  uncommon
4    common

In [96]:
traits.rarity.unique()

array(['common', 'uncommon', 'unique', 'rare'], dtype=object)

Mamy 4 typy
**może się przydać do wskazówek/if-ów**

*CR: Nie jestem pewna. Nie wiem czym są te wartości do końca i jaki mają wpływ.
Ale chyba jak się je zostawi i się nie przydadzą to nic się nie stanie.*

### Traits - senses

In [97]:
senses = unpack_column(traits, "senses")

In [98]:
senses.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   value   1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


In [99]:
senses.head()

value
0  darkvision, scent (imprecise) 60 feet, snow vi...
1        low-light vision, scent (imprecise) 30 feet
2        low-light vision, scent (imprecise) 60 feet
3              darkvision, scent (imprecise) 60 feet
4                                         darkvision

Jak na razie raczej nie dla nas - dane trochę "nieprzyjemnie" przechowywane (string z ",") // może się okazać, że gdzieś nie ma przecinka itp.

*Czy są ważne?*
// może dla ifów

*CR: Może dla ifów, ale to znowu w tą stronę że jak ktoś jest smokiem to widzi w ciemności, ale nie koniecznie w drugą stronę.*

*Nie do końca wiem jak to będzie działało*

### Traits - size

In [101]:
size = unpack_column(traits, "size")

In [102]:
size.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   value   1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


In [103]:
size.head()

value
0    lg
1   grg
2  huge
3   grg
4    sm

In [104]:
size.value.unique()

array(['lg', 'grg', 'huge', 'sm', 'med', 'tiny'], dtype=object)

*To chyba na CR się nie przekłada???*

//ale może być ewentualnie do ifów???


*CR: Ja bym to nawet uwzględniła, bo tak na oko jak coś jest duże to jest huge to będzie silne.*

* Najwyżej lasso nam wywali*

### Traits - value

In [105]:
value = unpack_column(traits, "value")

In [106]:
value.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       1206 non-null   object
 1   1       848 non-null    object
 2   2       258 non-null    object
 3   3       41 non-null     object
 4   4       3 non-null      object
dtypes: object(5)
memory usage: 56.5+ KB


In [107]:
traits.value

0                     [cold, dragon]
1                 [animal, dinosaur]
2                           [animal]
3               [amphibious, dragon]
4                [archon, celestial]
                   ...              
360                  [dragon, earth]
361                   [fey, gremlin]
362     [construct, golem, mindless]
363    [incorporeal, spirit, undead]
364              [caligni, humanoid]
Name: value, Length: 1206, dtype: object

In [109]:
unique_traits_value = defaultdict(lambda: 0)

for row in traits.value.values:
    for i in row:
        unique_traits_value[i] += 1

In [110]:
unique_traits_value.keys()

dict_keys(['cold', 'dragon', 'animal', 'dinosaur', 'amphibious', 'archon', 'celestial', 'aberration', 'azata', 'undead', 'gnome', 'humanoid', 'mindless', 'zombie', 'lizardfolk', 'construct', 'ghost', 'incorporeal', 'spirit', 'acid', 'fire', 'earth', 'elemental', 'aquatic', 'fey', 'hag', 'beast', 'angel', 'goblin', 'electricity', 'demon', 'fiend', 'xulgath', 'skeleton', 'human', 'werecreature', 'plant', 'water', 'tengu', 'soulbound', 'giant', 'monitor', 'protean', 'kobold', 'aeon', 'devil', 'leshy', 'vampire', 'psychopomp', 'mutant', 'gremlin', 'daemon', 'mummy', 'air', 'drow', 'elf', 'troll', 'genie', 'sea-devil', 'wraith', 'ooze', 'swarm', 'boggard', 'ghoul', 'inevitable', 'merfolk', 'nymph', 'fungus', 'alchemical', 'golem', 'duskwalker', 'orc', 'ratfolk', 'wight', 'dero', 'changeling', 'sprite', 'ethereal', 'duergar', 'dwarf', 'caligni', 'aasimar', 'rakshasa', 'gnoll', 'summoned', 'dhampir', 'catfolk', 'astral', 'tiefling', 'couatl', 'qlippoth', 'suli', 'petitioner', 'oni', 'serpentf

In [111]:
unique_traits_value

defaultdict(<function __main__.<lambda>()>,
            {'cold': 28,
             'dragon': 154,
             'animal': 205,
             'dinosaur': 12,
             'amphibious': 91,
             'archon': 9,
             'celestial': 37,
             'aberration': 81,
             'azata': 6,
             'undead': 92,
             'gnome': 6,
             'humanoid': 181,
             'mindless': 82,
             'zombie': 7,
             'lizardfolk': 3,
             'construct': 58,
             'ghost': 2,
             'incorporeal': 28,
             'spirit': 22,
             'acid': 7,
             'fire': 64,
             'earth': 57,
             'elemental': 84,
             'aquatic': 45,
             'fey': 61,
             'hag': 8,
             'beast': 94,
             'angel': 8,
             'goblin': 9,
             'electricity': 17,
             'demon': 18,
             'fiend': 99,
             'xulgath': 3,
             'skeleton': 8,
             'human': 24,


In [115]:
print(max(unique_traits_value.items(), key=lambda x: x[1]))

('animal', 205)


Dużo różnych wartości, ale najwięcej zwierzaczków
Analogicznie jak wyżej
*Może przyda się do ifów?*

*CR: Może się przyda, bo pewnie każda subrasa czy coś tam ma jakieś właściwości. Ale oby nie miała bo tego jest dużo. I z każdą książką prawdopodobnie będzie tego więcej!*
*Types są chyba nadrzędne na tym i oby to nam wystarczyło*

### Traits - attitiude

In [120]:
traits.attitude.head()

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: attitude, dtype: object

In [121]:
traits.attitude.info()

<class 'pandas.core.series.Series'>
Int64Index: 1206 entries, 0 to 364
Series name: attitude
Non-Null Count  Dtype 
--------------  ----- 
718 non-null    object
dtypes: object(1)
memory usage: 18.8+ KB


In [123]:
attitude = traits.attitude

In [124]:
attitude.info()

<class 'pandas.core.series.Series'>
Int64Index: 1206 entries, 0 to 364
Series name: attitude
Non-Null Count  Dtype 
--------------  ----- 
718 non-null    object
dtypes: object(1)
memory usage: 18.8+ KB


In [125]:
attitude.head()

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: attitude, dtype: object

In [128]:
attitude = attitude[pd.notnull(attitude)]

In [129]:
attitude.head()

9     {'value': 'hostile'}
10    {'value': 'hostile'}
15    {'value': 'hostile'}
27    {'value': 'hostile'}
37    {'value': 'hostile'}
Name: attitude, dtype: object

In [132]:
attitude = unpack_column(pd.DataFrame(attitude), "attitude")

In [133]:
attitude.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 718 entries, 9 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   value   718 non-null    object
dtypes: object(1)
memory usage: 11.2+ KB


In [134]:
attitude.head()

value
9   hostile
10  hostile
15  hostile
27  hostile
37  hostile

In [138]:
attitude.iloc[:, 0]

9      hostile
10     hostile
15     hostile
27     hostile
37     hostile
        ...   
358    hostile
359    hostile
361    hostile
363    hostile
364    hostile
Name: value, Length: 718, dtype: object

In [135]:
attitude.value.unique()

array(['hostile', 'indifferent'], dtype=object)

In [142]:
for u in attitude.value.unique():
    print(u)
    print(len([x for x in attitude.values if x == u]))

hostile
711
indifferent
7


My tylko 718 potworków ma attitude
* hostile: 711
* indifferent: 7

*Wydaje mi się, że może do if-ów się przydać*
*CR: Być może*

### Podsumowanie - traits
* languages - nie (99%)
* rarity - *do wskazówek(?)*
* senses - *może wskazówki????*
* size - *to chyba też raczej do wskazówek*
* value - *jw.*
* attitude - *jw.*

## Schema

In [143]:
schema = unpack_column(system, "schema")

In [144]:
schema.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   version        1206 non-null   float64
 1   lastMigration  0 non-null      float64
dtypes: float64(2)
memory usage: 28.3 KB


In [145]:
schema.head()

version  lastMigration
0    0.827            NaN
1    0.827            NaN
2    0.827            NaN
3    0.827            NaN
4    0.827            NaN

In [146]:
schema.version.unique()

array([0.827])

Wszystkie potworki mają te samą wartość (idk co reprezentuje)
*Raczej nie potrzebne*
*CR: ok*